src/Tutor.Infrastructure/Database/EventStore/IEventStore.cs

```diff
@@ -1,12 +1,13 @@
 ﻿using System;
 using System.Collections.Generic;
 using Tutor.Core.BuildingBlocks.EventSourcing;
+using Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries;
 
 namespace Tutor.Infrastructure.Database.EventStore
 {
     public interface IEventStore
     {
-        void Save(EventSourcedAggregateRoot aggregate);
+        void Save(KnowledgeComponentMastery aggregate);
         IEnumerable<DomainEvent> GetEvents(DateTime? start, DateTime? end);
```

Comment:
I definitely agree with changing the event hierarchy so all events we currently have inherit KnowledgeComponentEvent (I don't actually remember why it wasn't like that in the first place), but I'm not sure why the changes to the event store were necessary to support that? The purpose of the EventStore package and IEventStore especially was really just to serve as a generic facade to hide the technology used to implement the event store, with the idea that it could theoretically be completely extracted from SmartTutor without much changing, so tying it to the actual domain model changes the whole concept quite a bit.

Reply:
I wanted to create indexes on the LearnerId and KcId fields for the stored domain event. Considering that we are most interested in these types of events for now I found it fitting to repurpose the general DomainEvent.

The question is do we need to treat everything as the same base structure or can we have multiple documents (currently tables) for each major event type?

Reply:
If indexing is the goal, separate columns for LearnerId and KnowledgeComponentId aren't actually required, since Postgres is capable of indexing JSONB columns. Those indexes would have to be created using direct SQL though, since Npgsql doesn't have support for them. The main reason that approach wasn't used for TimeStamp is because time conversions in Postgres can't be used in indexes (and partially because time was assumed to be generic for every possible event).

As for whether or not we need to treat all events as the same base structure, it really depends on what we gain by that and what we lose.
When researching event sourcing, I didn't run into any approach that didn't treat events as the same base structure, because the goal is usually to keep them as structureless as possible. They need to have some metadata to make sure they can be associated to an aggregate and that they can be arranged chronologically, but for the actual payload, the idea is that the event store doesn't know or care what it contains. It's possible to arrange them into multiple collections (although I don't really see how that would work with a relational database and EF Core), but all the collections still have the same schema, and it's generally just an internal detail of the event store.

And for the approach used in these commits, where events don't have the same base structure / the event store actually cares about the structure of the payload instead of treating it as a blob, here are some consequences that I see:
- Adding a new aggregate becomes more difficult. Instead of just creating the aggregate with any events, and using the completely generic event store through a repository, it would be necessary to expand IEventStore (and all implementations) with specific methods for saving that aggregate. 
- Swapping database technologies for the event store becomes more difficult.
- Room for error when adding events to an existing aggregate. While it does make sense to have a one root event per aggregate and inherit from that, there's nothing actually forcing anyone adding new events to respect that. The Causes method as it is now will still accept any inheritor of DomainEvent, but the Save method would fail unexpectedly when it tries to cast it. If the approach we're going for is one root event and one collection per aggregate, EventSourcedAggregateRoot would probably need to become <code>EventSourcedAggregateRoot<T> where T : DomainEvent</code>.

So, what we would lose is the convenience that comes from the event store knowing nothing about the payload structure, which I would say is a pretty big loss, but I honestly have no idea how much we gain. Do we have any new information on what kind of queries we'll need the event store to perform, and how fast they'll actually need to be?